\hypertarget{ux6587ux6863ux6d4bux8bd5}{%
\subsection{文档测试}\label{ux6587ux6863ux6d4bux8bd5}}

如果你经常阅读 Python 的官方文档，可以看到很多文档都有示例代码。比如
\href{https://docs.python.org/3/library/re.html}{re
模块}就带了很多示例代码：

\begin{pythoncode}
>>> import re
>>> m = re.search('(?<=abc)def', 'abcdef')
>>> m.group(0)
'def'
\end{pythoncode}

可以把这些示例代码在 Python
的交互式环境下输入并执行，结果与文档中的示例代码显示的一致。

这些代码与其他说明可以写在注释中，然后，由一些工具来自动生成文档。既然这些代码本身就可以粘贴出来直接运行，那么，可不可以自动执行写在注释中的这些代码呢？

答案是肯定的。

当我们编写注释时，如果写上这样的注释：

\begin{pythoncode}
def abs(n):
    '''
    Function to get absolute value of number.
    
    Example:
    
    >>> abs(1)
    1
    >>> abs(-1)
    1
    >>> abs(0)
    0
    '''
    return n if n >= 0 else (-n)
\end{pythoncode}

无疑更明确地告诉函数的调用者该函数的期望输入和输出。

并且，Python 内置的
``文档测试''（doctest）模块可以直接提取注释中的代码并执行测试。

doctest 严格按照 Python
交互式命令行的输入和输出来判断测试结果是否正确。只有测试异常的时候，可以用\texttt{...}表示中间一大段烦人的输出。

让我们用 doctest 来测试上次编写的\texttt{Dict}类：

\begin{pythoncode}
class Dict(dict):
    '''
    Simple dict but also support access as x.y style.

    >>> d1 = Dict()
    >>> d1['x'] = 100
    >>> d1.x
    100
    >>> d1.y = 200
    >>> d1['y']
    200
    >>> d2 = Dict(a=1, b=2, c='3')
    >>> d2.c
    '3'
    >>> d2['empty']
    Traceback (most recent call last):
        ...
    KeyError: 'empty'
    >>> d2.empty
    Traceback (most recent call last):
        ...
    AttributeError: 'Dict' object has no attribute 'empty'
    '''
    def __init__(self, **kw):
        super(Dict, self).__init__(**kw)

    def __getattr__(self, key):
        try:
            return self[key]
        except KeyError:
            raise AttributeError(r"'Dict' object has no attribute '%s'" % key)

    def __setattr__(self, key, value):
        self[key] = value

if __name__=='__main__':
    import doctest
    doctest.testmod()
\end{pythoncode}

运行\texttt{python\ mydict2.py}：

\begin{pythoncode}
$ python mydict2.py
\end{pythoncode}

什么输出也没有。这说明我们编写的 doctest
运行都是正确的。如果程序有问题，比如把\texttt{\_\_getattr\_\_()}方法注释掉，再运行就会报错：

\begin{pythoncode}
$ python mydict2.py
**********************************************************************
File "/Users/michael/Github/learn-python3/samples/debug/mydict2.py", line 10, in __main__.Dict
Failed example:
    d1.x
Exception raised:
    Traceback (most recent call last):
      ...
    AttributeError: 'Dict' object has no attribute 'x'
**********************************************************************
File "/Users/michael/Github/learn-python3/samples/debug/mydict2.py", line 16, in __main__.Dict
Failed example:
    d2.c
Exception raised:
    Traceback (most recent call last):
      ...
    AttributeError: 'Dict' object has no attribute 'c'
**********************************************************************
1 items had failures:
   2 of   9 in __main__.Dict
***Test Failed*** 2 failures.
\end{pythoncode}

注意到最后 3 行代码。当模块正常导入时，doctest
不会被执行。只有在命令行直接运行时，才执行 doctest。所以，不必担心
doctest 会在非测试环境下执行。

\hypertarget{ux7ec3ux4e60}{%
\subsubsection{练习}\label{ux7ec3ux4e60}}

对函数\texttt{fact(n)}编写 doctest 并执行：

\begin{pythoncode}
# -*- coding: utf-8 -*-
\end{pythoncode}

\begin{pythoncode}
if __name__ == '__main__':
    import doctest
    doctest.testmod()
\end{pythoncode}

\hypertarget{ux5c0fux7ed3}{%
\subsubsection{小结}\label{ux5c0fux7ed3}}

doctest
非常有用，不但可以用来测试，还可以直接作为示例代码。通过某些文档生成工具，就可以自动把包含
doctest 的注释提取出来。用户看文档的时候，同时也看到了 doctest。

\hypertarget{ux53c2ux8003ux6e90ux7801}{%
\subsubsection{参考源码}\label{ux53c2ux8003ux6e90ux7801}}

\href{https://github.com/michaelliao/learn-python3/blob/master/samples/debug/mydict2.py}{mydict2.py}

